client/comms/wsconn_test.go

```diff
@@ -0,0 +1,463 @@
+	}
+
+	log.Infof("Done generating TLS certificates")
+	return nil
+}
+
+func TestReconnect(t *testing.T) {
+	upgrader := websocket.Upgrader{}
+
+	pingCh := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
+
+	pingWait := time.Millisecond * 200
+	writeWait := time.Millisecond * 200
+
+	var wsc *WsConn
+
+	id := uint64(0)
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		hCtx, hCancel := context.WithCancel(context.Background())
+		atomic.AddUint64(&id, 1)
+
+		c, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			t.Fatalf("unable to upgrade http connection: %s", err)
+		}
+
+		c.SetPongHandler(func(string) error {
+			id := atomic.LoadUint64(&id)
+			t.Logf("handler #%d: pong received", id)
+			return nil
+		})
+
+		go func() {
+			for {
+				select {
+				case <-pingCh:
+					err := c.WriteControl(websocket.PingMessage, []byte{},
+						time.Now().Add(writeWait))
+					if err != nil {
+						t.Errorf("handler #%d: ping error: %v", id, err)
+						return
+					}
+
+					id := atomic.LoadUint64(&id)
+					t.Logf("handler #%d: ping sent", id)
+
+				case <-hCtx.Done():
+					return
+				}
+			}
+		}()
+
+		for {
+			_, _, err := c.ReadMessage()
+			if err != nil {
+				c.Close()
+				hCancel()
+
+				if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+					// Terminate on a normal close message.
+					return
+				}
+
+				t.Errorf("handler #%d: read error: %v", id, err)
+				return
+			}
+
+		}
+	}
+
+	certFile, err := ioutil.TempFile("", "certfile")
+	if err != nil {
+		t.Fatalf("unable to create temp certfile: %s", err)
+	}
+	certFile.Close()
+	defer os.Remove(certFile.Name())
+
+	keyFile, err := ioutil.TempFile("", "keyfile")
+	if err != nil {
+		t.Fatalf("unable to create temp keyfile: %s", err)
+	}
+	keyFile.Close()
+	defer os.Remove(keyFile.Name())
+
+	err = genCertPair(certFile.Name(), keyFile.Name(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	host := "127.0.0.1:6060"
+	mux := http.NewServeMux()
+	mux.HandleFunc("/ws", handler)
+
+	server := &http.Server{
+		WriteTimeout: time.Second * 5,
+		ReadTimeout:  time.Second * 5,
+		IdleTimeout:  time.Second * 5,
+		Addr:         host,
+		Handler:      mux,
+	}
+
+	defer server.Close()
+
+	go func() {
+		err = server.ListenAndServeTLS(certFile.Name(), keyFile.Name())
+		if err != nil {
+			t.Log(err)
+		}
+	}()
+
+	cfg := &WsCfg{
+		Host:      host,
+		Path:      "ws",
+		PingWait:  pingWait,
+		WriteWait: writeWait,
+		RpcCert:   certFile.Name(),
+		Ctx:       ctx,
+		Cancel:    cancel,
+	}
+	wsc, err = NewWsConn(cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for idx := 0; idx < 5; idx++ {
+		// Drop the connection and force a reconnect by waiting.
+		time.Sleep(time.Millisecond * 210)
+
+		// Wait for a reconnection.
+		for !wsc.isConnected() {
+			time.Sleep(time.Millisecond * 10)
+			continue
+		}
+
+		// Send a ping.
+		pingCh <- struct{}{}
+	}
+
+	time.Sleep(time.Millisecond * 200)
+	cancel()
+	time.Sleep(time.Millisecond * 200)
```

Comment:
These `Sleep`s are a red flag that a shutdown synchronization mechanism is needed.  How would the client process shutdown cleanly?

Reply:
noted, switched to using a waitgroup to block and wait for shutdown to conclude before returning.